File: Aman/Trie/Implement Trie (Prefix Tree).cpp
class Trie {
public:
    
    struct TrieNode{
        struct TrieNode* child[26];
        bool isEnd;
    };
    TrieNode* node;
    
    /** Initialize your data structure here. */
    Trie() {
      node=new TrieNode();  
    }
    
    /** Inserts a word into the trie. */
    void insert(string word) {
        TrieNode* cur=node;
        for(char c: word)
        {
            if(cur->child[c-'a']==NULL)
              cur->child[c-'a']=new TrieNode();
            cur=cur->child[c-'a'];
        }
        cur->isEnd=true;
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) {
        TrieNode* cur=node;
        for(char c: word)
        {
            if(cur->child[c-'a']==NULL)
                return false;
            cur=cur->child[c-'a'];   
        }
        if(cur->isEnd==true)
            return true;
        return false;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        TrieNode* cur=node;
        for(char c: prefix)
        {
            if(cur->child[c-'a']==NULL)
                return false;
            cur=cur->child[c-'a'];
        }
        return true;
    }
};

File: Harish Kumar/Trie/Implement Trie (Prefix Tree)/Implement Trie (Prefix Tree).java
class TrieNode {
    
    TrieNode map [];
    boolean isEnd;
    public TrieNode(){
        
        map = new TrieNode[26];
        isEnd = false;
    }
    
}

class Trie {
    /** Initialize your data structure here. */
    TrieNode root;
    public Trie() {
        
        root = new TrieNode();
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        TrieNode temp = root;
        for(char c : word.toCharArray())
        {
            if(temp.map[c - 'a'] == null)
            {
                temp.map[c - 'a'] = new TrieNode();
            }
            temp = temp.map[c - 'a'];
        }
        
        temp.isEnd = true;
        
    }
    
    public TrieNode commonSearch(String word)
    {
        TrieNode temp = root;
        
        for(char c : word.toCharArray())
        {
            if(temp.map[c - 'a'] == null)
            {
                return null;
            }
            
            temp = temp.map[c - 'a'];
        }
        return temp;
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        
        TrieNode temp = commonSearch(word);
        return temp != null && temp.isEnd;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        
        return commonSearch(prefix) != null;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */
